### PR TITLE
✨ use goreleaser for releasing binary

### PR DESCRIPTION
**Pull request by kranurag7**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

For automating releases. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4391 https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4091

Related: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4610 https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4309

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Use goreleaser for releasing binary
```

### CODE REVIEW

File: .github/workflows/release.yaml
Comment (richardcase) [inline .github/workflows/release.yaml:6]: Currently we sometimes tag main and then create a release branch. We should think about whether we stop this as part of the release.

File: .goreleaser.yaml
Comment (richardcase) [inline .goreleaser.yaml:4]: We also need to build a `clusterawsadm` binary.

File: cmd/clusterawsadm/cmd/version/version.go
Comment (richardcase) [inline cmd/clusterawsadm/cmd/version/version.go:66]: We will need to get the binary name to cover both binaries we build.

**Review by richardcase — COMMENTED** [review]
Thanks for this @kranurag7 . I like the use of goreleaser here and creating the draft release.

File: .github/workflows/release.yaml
Comment (kranurag7) [reply .github/workflows/release.yaml]: I switched to using `v*` for this workflow to trigger.

**Comment by nrb** [comment]
This lgtm, but I am not familiar with the release process and therefore don't feel comfortable putting an actual review on this.

**Comment by dlipovetsky** [comment]
From today's officeh ours: We'd like to have more maintainers/reviewers familiar with the current release process before changing this.

/milestone v2.5.0

**Comment by vincepri** [comment]
this seems fine to me although I never used goreleaser, for comparison this is what upstream does https://github.com/kubernetes-sigs/cluster-api/blob/main/.github/workflows/release.yaml

Let's wait to cut v2.4.0 then merge this for subsequent releases?

**Comment by nrb** [comment]
After going through the v2.4.1 release process, we ran into #4881. Using goreleaser bypasses this.

I tested this with the following process:

```shell
gh pr checkout 4748
goreleaser release --snapshot --clean
```

It created all the binaries faster than the existing Docker method did, along with tarballs including `clusterctl-aws` and `clusterawsadm` binaries for every platform.

I'd like for Damiano to take a look at this too, but I think this is worth setting up for at least creating the release and uploading binaries.

We can look at how the release notes and images get built later. I'm guessing we could roll the release notes process into this, but maybe not the image building/promotion.

/lgtm

/assign @damdo

File: .github/workflows/release.yaml
Comment (nrb) [reply .github/workflows/release.yaml]: I like this solution - we can still do the release branch and build image while the binaries are building after the push.

**Comment by kranurag7** [comment]
Hey there, 
Happy to work on this till completion. Please ask if there's any question on the PR. I'll be responsive :) 

I rebased the PR and I created a demo release on my fork.
https://github.com/kranurag7/cluster-api-provider-aws/releases/tag/v2.4.2 

Other things like removing binaries existing binaries related make targets, updating release docs etc. can be done in a subsequent PR or even this PR as well.

**Comment by nrb** [comment]
> Happy to work on this till completion.

Thank you!

> Other things like removing binaries existing binaries related make targets

Let's not do that quite yet, IMO. I think once we get a release or two out with goreleaser we can do it, though.

> updating release docs

That would be great! We've been using [this doc](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/book/src/development/releasing.md).

File: .github/workflows/release.yaml
Comment (nrb) [inline .github/workflows/release.yaml:31]: @richardcase I think a maintainer would need to create a GitHub token for the goreleaser job to use. I do not have access to the repo secrets.

Comment (nrb) [reply .github/workflows/release.yaml:31]: Ah, good call. Thanks!

**Comment by nrb** [comment]
/lgtm

I think we should merge this and iterate on the docs.

What do you think @vincepri and @richardcase? I know it's not the same as the upstream CAPI process, but I do feel like using goreleaser is less likely to run into filesystem mounting issues between mac/Linux when run locally.

File: .goreleaser.yaml
Comment (damdo) [inline .goreleaser.yaml:51]: What does ignore do here?
It looks like we are releasing a windows for arm64 at the moment.
[clusterawsadm-windows-arm64.exe](https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/download/v2.4.1/clusterawsadm-windows-arm64.exe)

Comment (damdo) [reply .goreleaser.yaml:51]: cc. @kranurag7

Comment (kranurag7) [reply .goreleaser.yaml:51]: I think I've never seen windows running on arm. If something like that exists 🙂, I can update it. 
Also, I'm unsure if we also need to support freebsd here.

Comment (nrb) [reply .goreleaser.yaml:51]: > I think I've never seen windows running on arm.

It's rare, but Microsoft made a couple Surface laptops that support ARM, as well as a RasPi image. See https://learn.microsoft.com/en-us/windows/arm/overview. Since we're already shipping it, probably worth continuing.

I don't think supporting BSD is a priority. We do not currently build for it.

Comment (damdo) [reply .goreleaser.yaml:51]: +1 to what @nrb said.
I'd personally re-add windows arm64, and leave freeBSD out.
Thanks!

Comment (nrb) [reply .goreleaser.yaml:48]: Looks like it's still ignoring windows-arm

Comment (kranurag7) [reply .goreleaser.yaml]: We will be releasing two binaries `clusterawsadm` and `clusterctl-aws`
I removed it only from the first binary and missed it in the second one so I updated the same.  

The other thing I updated is following given goreleaser new release includes this. 
```diff
-  - -X 'sigs.k8s.io/cluster-api-provider-aws/v2/version.gitTreeState={{ if .IsGitDirty }}dirty{{ else }}clean{{ end }}'
+  - -X 'sigs.k8s.io/cluster-api-operator/version.gitTreeState={{.GitTreeState}}'
```

**Review by richardcase — CHANGES_REQUESTED** [review]
Thanks for this @kranurag7 . As part of this i think we also need to update the makefile as we currently create a release:
- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/Makefile#L506

And also we need to update the release process docs:
- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/book/src/development/releasing.md

**Comment by richardcase** [comment]
It would also be good to initially disable the change log:

```
changelog:
  disable: true
```

**Comment by nrb** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4748/pull-cluster-api-provider-aws-build-docker/1773297258993815552#1:build-log.txt%3A530

>  make: *** No rule to make target 'release-binaries'.  Stop. 

Could we rewrite this target to invoke goreleaser instead?

File: .goreleaser.yaml
Comment (nrb) [inline .goreleaser.yaml:64]:
```suggestion
  disable: true

release:
  discussion_category_name: General

  extra_files:
    - glob: ./templates/*.yaml

  draft: true
```

This suggestion comes from examining the contents of https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/tag/v2.4.1. Mostly, it's including the template files which are fetched by clusterctl.

It also moves the choice to be a draft into the configuration file vs a command line argument.

File: Makefile
Comment (nrb) [inline Makefile:506]:
```suggestion
	$(GH) release create $(VERSION) --notes-file $(RELEASE_DIR)/CHANGELOG.md --title $(VERSION) --repo $(GH_REPO)
```

Move the draft option into `.goreleaser.yaml`; see comment on that file.

**Comment by nrb** [comment]
@kranurag7 I made a [commit](https://github.com/nrb/cluster-api-provider-aws/commit/b9367fdfd1a21d84280bfe23b790305ad80e666a) which I believe will fix the CI job, as well as align what goreleaser produces more closely to the current contents of a release.

This includes the suggestions I've made so far. You're welcome to grab it and incorporate it.

**Comment by kranurag7** [comment]
Thank you, I think we should also decide how templates gets uploaded. As of now, we do it via `make` target. 

If we intend to do it via goreleaser then the corresponding make targets needs to be adjusted accordingly.

```make
.PHONY: release-templates
release-templates: $(RELEASE_DIR) ## Generate release templates
	cp templates/cluster-template*.yaml $(RELEASE_DIR)/

.PHONY: upload-gh-artifacts
upload-gh-artifacts: $(GH) ## Upload artifacts to Github release
	$(GH) release upload $(VERSION) -R $(GH_REPO) --clobber  $(RELEASE_DIR)/*
```

NOTE: This is not a blocker because we are using `--clobber` meaning that we are overwriting existing assets with the same name. But it would be good to have it only once.

**Comment by nrb** [comment]
> If we intend to do it via goreleaser then the corresponding make targets needs to be adjusted accordingly.

Pushed a commit that removes these targets, as well as removing them from the documentation that we've used in the past.

@faiq @richardcase @damdo would be helpful if you double checked the docs changes I've made.

File: docs/book/src/development/releasing.md
Comment (nrb) [reply docs/book/src/development/releasing.md:59]: Also need to retain the `make release-manifests` output to get our `metadata.yaml` file updated and uploaded

File: .goreleaser.yaml
Comment (nrb) [inline .goreleaser.yaml:61]: We can drop this format override since we're uploading binaries now, not archives.

File: .goreleaser.yaml
Comment (nrb) [inline .goreleaser.yaml:71]: Add `metadata.yaml` as an uploaded artifact

**Comment by nrb** [comment]
/hold

Holding because there are some missing artifacts when releasing with this configuration.

File: Makefile
Comment (nrb) [inline Makefile:603]: Un-deprecate and rename this so we know it's for CI/dry runs of the build.

File: .goreleaser.yaml
Comment (nrb) [inline .goreleaser.yaml:1]: Note for @richardcase since you asked about it before - the `changelog.disabled` flag actually disabled posting a changelog/release notes entirely; even customized ones passed via `--release-notes`. The `--release-notes` flag is what overrides the Goreleaser changelog generation.

See https://goreleaser.com/customization/changelog/

**Review by damdo — COMMENTED** [review]
/lgtm

/assign @richardcase

File: .goreleaser.yaml
Comment (richardcase) [reply .goreleaser.yaml:1]: Noted and understood, thanks @nrb . I mentioned it more in the context that currently the makefile would generate and attach the release notes. But if we are attaching the custom release notes directly with goreleaser then great :+1:

**Comment by richardcase** [comment]
Great work @kranurag7 and @nrb , this is a real quality of life improvement :tada: 

/approve

**Comment by nrb** [comment]
/cherry-pick release-2.4

The `.goreleaser.yaml` file needs to be present in the branch when its checked out during the build.

We worked around this for v2.4.2, but the cherry-pick will be useful for releases going forward.

**Pull request merged**
